cmake: support snippets scope for `zephyr_get()`

tejlmand opened the pull request:
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/57139

Snippets provides the possibility of defining EXTRA_DTC_OVERLAY_FILE and
EXTRA_CONF_FILE in snippets.
Snippets must co-exist with existing infrastructure of
EXTRA_DTC_OVERLAY_FILE and EXTRA_CONF_FILE, and a user specifying a
snippet must be able to specify extra files for adjusting the snippet.

This means that if the following is specified:
`-DSNIPPET=some_snippet -DEXTRA_CONF_FILE=extra.conf`
then `extra.conf` may contain adjustments to the snippet.
Similar to sysbuild. Imagine a sysbuild controlled image uses a default
snippet for building, by ensuring that any extra
`-D<image>_EXTRA_CONF_FILE=extra.conf` arguments takes precedence over
the snippet we allow users to make adjustments if they need.

This commit introduces a snippets scope where snippet scoped variables
can be set with `zephyr_set()` and then `zephyr_get()` will take the
snippet scoped variables into consideration before returning.

Adjust calls to `zephyr_get(EXTRA_DTC_OVERLAY_FILE)` and
`zephyr_get(EXTRA_CONF_FILE)` to use `MERGE` to ensure all scopes are
considered.

tejlmand committed:
cmake: add reverse option to `zephyr_get(... MERGE)`

Adding the possibility to have the list returned in reversed order when
using `zephyr_get(... MERGE).

`zephyr_get(... MERGE)` creates a list which populates the content based
on variable settings in the following scopes, in this order:
sysbuild, cache, environment, local.

This works well for lists where content first in list has highest
precedence, such as ROOTs settings.
However, for settings where the value last in the list will overwrite
values earlier in the list, we want the list to be reversed, examples
of such can be CONF_FILE, OVERLAY_CONFIG, DTC_OVERLAY_FILE, where the
content of the file last in the list will overrule the content from an
earlier file.
So to ensure that a DTC_OVERLAY_FILE defined as cache takes precedence
over an env or local scope variable the possibility of reversing the
list must be available.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: support snippets scope for `zephyr_get()`

Fixes: #57139

Snippets provides the possibility of defining EXTRA_DTC_OVERLAY_FILE and
EXTRA_CONF_FILE in snippets.
Snippets must co-exist with existing infrastructure of
EXTRA_DTC_OVERLAY_FILE and EXTRA_CONF_FILE, and a user specifying a
snippet must be able to specify extra files for adjusting the snippet.

This means that if the following is specified:
`-DSNIPPET=some_snippet -DEXTRA_CONF_FILE=extra.conf`
then `extra.conf` may contain adjustments to the snippet.
Similar to sysbuild. Imagine a sysbuild controlled image uses a default
snippet for building, by ensuring that any extra
`-D<image>_EXTRA_CONF_FILE=extra.conf` arguments takes precedence over
the snippet we allow users to make adjustments if they need.

This commit introduces a snippets scope where snippet scoped variables
can be set with `zephyr_set()` and then `zephyr_get()` will take the
snippet scoped variables into consideration before returning.

Adjust calls to `zephyr_get(EXTRA_DTC_OVERLAY_FILE)` and
`zephyr_get(EXTRA_CONF_FILE)` to use `MERGE` to ensure all scopes are
considered.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented on cmake/modules/snippets.cmake:
```diff
@@ -93,14 +93,11 @@ function(zephyr_process_snippets)
   # separate command because build system files are not fond of
   # newlines.
-  list(TRANSFORM SNIPPET_NAMES PREPEND "COMMAND;${CMAKE_COMMAND};-E;echo;"
+  get_target_property(snippet_names snippets_target NAMES)
+  list(TRANSFORM snippet_names PREPEND "COMMAND;${CMAKE_COMMAND};-E;echo;"
     OUTPUT_VARIABLE snippets_target_cmd)
   add_custom_target(snippets ${snippets_target_cmd} USES_TERMINAL)
```
@mbolivar-nordic I considered to have just a single `snippets` target for this and then use the same target for scoping the variables as for the print command.
But because the printing itself has been constructed so that each snippet name requires `COMMAND` in front, then we need to fetch the list prior to creating the custom target. (we cannot a generator expression in this case).

If the `snippets` target itself is just for debugging, then just printing a space separated list of snippets would make it possible to simplify the code, but before proposing so, then I would like to know if there are special requirements for the `snippets` targets so that it must use newlines.

mbolivar-nordic replied:
This is done for consistency with boards and shields. I expect the number of snippets may grow too large over time to comfortably fit on one line.

tejlmand replied:
thanks for the explanation.

keeping as-is for now, but will consider to rework into a dedicated call to `snippets.py` when `snippets` target is specifically invoked.
That will be even more consistent with what we do for boards, and remove the need for handling the list here, thus making it possible to make cleaner CMake code.

stephanosio reviewed (changes requested):
This does not seem to achieve the same effect as https://github.com/zephyrproject-rtos/zephyr/pull/57513, which ensures that the snippet overlays are processed _after_ `DTC_OVERLAY_FILE` and `OVERLAY_CONFIG`.

From #57513:

> The snippet Kconfig and devicetree overlays are processed after the overlays specified in the OVERLAY_CONFIG and DTC_OVERLAY_FILE variables in order to ensure that the snippet configurations take precedence over the application default configurations -- this is necessary because snippets can be used to further customise and override the application default configurations.

I have a set of tests to verify the above behaviour in https://github.com/stephanosio/zephyr/commits/snippets_proc_order_tejlmand, with your patches here.


```bash
$ scripts/twister -v -N -M -p native_posix -T tests/cmake/snippets
...
INFO    - 1/5 native_posix              tests/cmake/snippets/cmake.snippets.bar             FAILED Failed (native 0.011s)
ERROR   - see: /home/stephanos/Dev/zephyrproject/zephyr/twister-out/native_posix/tests/cmake/snippets/cmake.snippets.bar/handler.log
INFO    - 2/5 native_posix              tests/cmake/snippets/cmake.snippets.none           PASSED (native 0.010s)
INFO    - 3/5 native_posix              tests/cmake/snippets/cmake.snippets.foo_bar         FAILED Failed (native 0.009s)
ERROR   - see: /home/stephanos/Dev/zephyrproject/zephyr/twister-out/native_posix/tests/cmake/snippets/cmake.snippets.foo_bar/handler.log
INFO    - 4/5 native_posix              tests/cmake/snippets/cmake.snippets.bar_foo         FAILED Failed (native 0.008s)
ERROR   - see: /home/stephanos/Dev/zephyrproject/zephyr/twister-out/native_posix/tests/cmake/snippets/cmake.snippets.bar_foo/handler.log
INFO    - 5/5 native_posix              tests/cmake/snippets/cmake.snippets.foo             FAILED Failed (native 0.006s)
ERROR   - see: /home/stephanos/Dev/zephyrproject/zephyr/twister-out/native_posix/tests/cmake/snippets/cmake.snippets.foo/handler.log
```

Note that the above tests pass with #57513.

tejlmand commented:
>  which ensures that the snippet overlays are processed after DTC_OVERLAY_FILE and OVERLAY_CONFIG.

and that is what I would argue that we don't want, because a user should be able to specify a common snippet and from there specify **additional** adjustments in a DTC_OVERLAY_FILE` or more likely in an `OVERLAY_CONFIG`.

If snippets takes precedence over OVERLAY_CONFIG, then user must completely abandon the specific snippet / create a custom snippet if adjustments are needed.

stephanosio commented:
> > which ensures that the snippet overlays are processed after DTC_OVERLAY_FILE and OVERLAY_CONFIG.
> 
> and that is what I would argue that we don't want, because a user should be able to specify a common snippet and from there specify **additional** adjustments in a `DTC_OVERLAY_FILE` or more likely in an `OVERLAY_CONFIG`.
> 
> If snippets takes precedence over OVERLAY_CONFIG, then user must completely abandon the specific snippet / create a custom snippet if adjustments are needed.

There is no need for them to abandon the specific snippet; they just need to create an additional snippet with the extra configs to be applied on top of the existing snippet(s), which is not so much different from creating additional Kconfig and devicetree overlays.

Nevertheless, I can see that forcing users to use snippets for this purpose can be less than ideal.

What I need is to ensure that the `prj.conf` and `app.overlay` to get processed _before_ the snippet overlays so that snippets can override the "application defaults" specified in those files.

For example, if `app.overlay` contains `deleted-by-foo` node, a snippet should be able to delete that `deleted-by-foo` node (which is part of the things that https://github.com/stephanosio/zephyr/commit/a21dcf66035b2cc303e3170184622d55b0d68cac tests).

That is currently not the case with the patches in this PR.

tejlmand commented:
> > For example, if app.overlay contains deleted-by-foo node, a snippet should be able to delete that deleted-by-foo node (which is part of the things that [stephanosio@a21dcf6](https://github.com/stephanosio/zephyr/commit/a21dcf66035b2cc303e3170184622d55b0d68cac) tests).

> I agree with this.

If you read existing documentation, then applying `DTC_OVERLAY_FILE` will disable any automatic sourcing of overlay files.
See [Set devicetree overlays](https://docs.zephyrproject.org/latest/build/dts/howtos.html#set-devicetree-overlays).

This means that as soon that a snippet is used, and that snippet sets `DTC_OVERLAY_FILE`, then `app.overlay` is no longer sourced.
@stephanosio are you proposing to change that behavior ?

Because if you are, then that requires a more changes to our documentation as to when `<board>.overlay` / `app.overlay` are automatically sourced.

`prj.conf` (CONF_FILE) is a bit different, because `prj.conf` will only be replaced if `CONF_FILE` is specified, but snippets uses `OVERLAY_CONFIG` which is always used in combination with `CONF_FILE` and not replacing it.
`prj.conf` is always first in list, and therefore **any** `OVERLAY_CONFIG` will (regardless of scope) will have the ability to override settings in `prj.conf`
Ref: [the initial configuration](https://docs.zephyrproject.org/latest/build/kconfig/setting.html#the-initial-configuration)

tejlmand commented:
>  (e.g. SNIPPET_DTC_OVERLAY_FILE and SNIPPET_OVERLAY_CONFIG) and they work independently of the DTC_OVERLAY_FILE and OVERLAY_CONFIG CMake variables
> ....
> If we do that, snippets no longer need to comply with the documented behavioural characteristics of DTC_OVERLAY_FILE and OVERLAY_CONFIG (which is quite unnecessary, if I may add).

which give even more variations of how to configure the build system.
Also we need to careful describe how those settings behaves, and ensure they behave consistently.

The part I like wrt. current snippets implementation is that it works as an extension to current design and reuse same CMake variables names, meaning users are already custom to what those settings do.
Similar, it will be easier to allow other CMake variables to be set in snippets, in case we find need for that.

One thing, which has actually been raised before, we could consider, is to have a setting similar to `OVERLAY_CONFIG`.
Today we have:
- CONF_FILE for config files
- DTC_OVERLAY_FILE for dtc overlay, a setting which behavior is similar to `CONF_FILE` in the way it disable automatic sourcing.


- CONFIG_OVERLAY for config files, where this setting is used in combination with CONF_FILE.

Devicetree overlays has no comparable setting to `CONFIG_OVERLAY`, so maybe it would be better to align how config files and overlay files works and can be specified.
This could be achieved by having a general setting for additional overlay files, similar to how we have `CONFIG_OVERLAY` for configs.
That would allow snippets to make use of such setting, and still ensure that what a snippet does can actually be done manually by hand.

tejlmand commented:
> This could be achieved by having a general setting for additional overlay files, similar to how we have CONFIG_OVERLAY for configs.
That would allow snippets to make use of such setting, and still ensure that what a snippet does can actually be done manually by hand.

The name `OVERLAY_CONFIG` sometimes causes confusion with `DTC_OVERLAY_FILE`, both in what they configure and how they behave.

Maybe a solution could be to deprecate `OVERLAY_CONFIG` and instead introduce `EXTRA_CONF_FILE`.

That way we could have:
- `DTC_OVERLAY_FILE`, and `EXTRA_DTC_OVERLAY_FILE`
- `CONF_FILE`, and `EXTRA_CONF_FILE`

which resembles what we have with `ZEPHYR_MODULES`, like
- `ZEPHYR_MODULES` and `ZEPHYR_EXTRA_MODULES` (should then probably be renamed to `EXTRA_ZEPHYR_MODULES`.

this will bring more consistency to how build variables are named and used, as well as being more intuitive in how they behave.

`<var-name>` replaces any automatic / default assignment.
`EXTRA_<var-name>` extends `<var-name>` with extra configuration.

stephanosio commented:
> which give even more variations of how to configure the build system.

`SNIPPET_DTC_OVERLAY_FILE` and `SNIPPET_OVERLAY_CONFIG` (or whatever we decide to call them) will be internal to the build system, so introducing them should not result in any increased complexity from the user's perspective. From the build system's perspective, they provide more clarity since they establish a clear boundary between the snippet configs and the rest.

> The part I like wrt. current snippets implementation is that it works as an extension to current design and reuse same CMake variables names, meaning users are already custom to what those settings do.

But that also results in conflicts with the existing infrastructure that utilise and modify these variables, as we have already seen.

> Similar, it will be easier to allow other CMake variables to be set in snippets, in case we find need for that

But do we really want that behaviour? Doing so, once again, has a lot of potential for conflicts with the existing infrastructure and makes it difficult to clearly define and enforce the snippet-specific behaviours (e.g. some other module decides to touch the CMake variable and all hell breaks loose).

I am hoping to end up with something more self contained and well isolated from the rest of the build system parts so that I do not have worry about breaking something completely unrelated when I make changes to the snippet implementation (or vice versa); generally speaking, writing self contained modules and reducing the number of global variables that get both modified processed by multiple modules would be good practice (though, that does not seem to be a very common thing when it comes to CMake ....).

stephanosio commented:
> The name `OVERLAY_CONFIG` sometimes causes confusion with `DTC_OVERLAY_FILE`, both in what they configure and how they behave.
> 
> Maybe a solution could be to deprecate `OVERLAY_CONFIG` and instead introduce `EXTRA_CONF_FILE`.
> 
> That way we could have:
> 
> * `DTC_OVERLAY_FILE`, and `EXTRA_DTC_OVERLAY_FILE`
> * `CONF_FILE`, and `EXTRA_CONF_FILE`

That does sound like a major improvement in terms of consistency, and I agree it would be good to implement that regardless of whether we decide to make the snippet implementation use them.

I still have concerns about the snippets directly modifying the global CMake variables though as mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/57661#issuecomment-1542569796.

stephanosio commented:
Also another open question is:

> The part I like wrt. current snippets implementation is that it works as an extension to current design and reuse same CMake variables names, meaning users are already custom to what those settings do.

If we decide to go down this path, how would we go about the defconfig support (and even something like `Kconfig.snippet` if we ever decide it is necessary), which is not currently handled through any particular global CMake variables.

We could create an exception for defconfigs only and give it a special treatment; but, that would be an inconsistency.

tejlmand commented:
>  should not result in any increased complexity from the user's perspective. 

well, if we inside snippets call them `OVERLAY_CONFIG`, then they indicate they behave the same, which they suddenly may not do.
If we call them `SNIPPETS_OVERLAY_CONFIG`, then it's added complexity in their naming.

>  as we have already seen.

what we have seen is that we need a functionality similar to `OVERLAY_CONFIG`  but for devicetree overlays.
This is something we don't have today. If we were to provide such a feature only for snippets, then that would be introducing inconsistency in the form as `yes, you can replace devicetree overlay from command line, but if you want to append to devicetree overlay you must create a snippet to do so.`

Therefore such functionality should be available both for snippets but also generally.

> But do we really want that behaviour?  Doing so, once again, has a lot of potential for conflicts with the existing infrastructure and makes it difficult to clearly define and enforce the snippet-specific behaviours 

Do we want to create an infrastructure where such extensibility would not be possible ?
Notice how `zephyr_get()` allows us to control which scopes takes precedence, thus a CMake module which set FOO var in local scope will take lower precedence than the value defined in sysbuild, cache, or snippet scope.

With your current suggestion, you would end up in:
1. snippets before `zephyr_get()`, like:
    ```
    zephyr_get(FOO ...)
    some_function(
      ${SNIPPETS_FOO}  # <-- Snippets version of FOO, not returned by zephyr_get in your proposal
      ${FOO}           # <-- The return result of zephyr_get(FOO)
    )
    ```
2. snippets after `zephyr_get()`, like:
    ```
    zephyr_get(FOO ...)
    some_function(
      ${FOO}           # <-- The return result of zephyr_get(FOO), not returned by zephyr_get in your proposal
      ${SNIPPETS_FOO}  # <-- Snippets version of FOO
    )
    ```

This bears the risk that sometimes snippets var take precedence over the same var in other scopes, and sometimes it doesn't.
If the snippets var is scoped independently (as this PR proposes) then the value from snippets are part of `zephyr_get()` and we ensure that snippet defined values always takes the same precedence, compared to other scopes.

> (e.g. some other module decides to touch the CMake variable and all hell breaks loose).

No other CMake module can touch the snippet variable, cause snippets variables and their values are scoped using a custom snippet target, and `zephyr_get()` controls which scopes takes precedence when returning the value(s).

----------------------

and one more advantage with the `zephyr_get()` is that we can have a single location where order of precedence is described.

tejlmand commented:
> how would we go about the defconfig support (and even something like Kconfig.snippet if we ever decide it is necessary), which is not currently handled through any particular global CMake variables.

There is not necessarily a direct problem in certain functionality only available through snippets.
Just like Zephyr modules may provide extra information wrt. twister tests / samples available.
But we should of course careful consider if a functionality provided through snippets would make sense to have as a general functionality.

So for a defconfig / Kconfig.snippets we should definitely consider if such functionality should be more generally available.
An example where a more general functionality could make sense is Zephyr modules.

Currently all Zephyr modules Kconfigs are sourced through this:
https://github.com/zephyrproject-rtos/zephyr/blob/90d160e689a170a2324444660badcaf10c953e3b/Kconfig.zephyr#L37-L39
https://github.com/zephyrproject-rtos/zephyr/blob/7c228c90428f6635a2e3caed85c492677594746a/modules/Kconfig#L13

meaning all potential Zephyr module defconfigs are sourced as part of the Zephyr module's Kconfig.
But if we made the snippets defconfig more generic, then it could potentially be utilized by Zephyr modules, and that in return would allow us to split Zephyr modules defconfigs from rest of their Kconfig.

Currently our Zephyr module Kconfig tree would be very vulnerable if code snippets linked above would be moved around.

mbolivar-nordic commented:
FTR I think the EXTRA_DTC_OVERLAY_FILE and EXTRA_CONF_FILE ideas make great sense and would be nice to get in before the release so we can move snippets over to them exclusively. Having snippet-specific versions of these variables seems unnecessary IMO as long as the order that snippets happen in is well defined (i.e. snippets append first, then CMake cache variables are appended on after would make sense to me).

Regarding Kconfigs/defconfigs, the reason I put the CMake variables under an `append:` was to be able to create sibling keys that would do things that have no direct CMake variables controlling them. Something like this:

```yaml
name: foo
append:
  EXTRA_DTC_OVERLAY_FILE: my.overlay
defconfig: my_defconfig
```

Where we pick names like `defconfig`, etc. that should be reasonably self documenting about what their role is (and of course document them as well)

Thoughts?

stephanosio commented:
> FTR I think the EXTRA_DTC_OVERLAY_FILE and EXTRA_CONF_FILE ideas make great sense and would be nice to get in before the release so we can move snippets over to them exclusively. Having snippet-specific versions of these variables seems unnecessary IMO as long as the order that snippets happen in is well defined (i.e. snippets append first, then CMake cache variables are appended on after would make sense to me).

Based on https://github.com/zephyrproject-rtos/zephyr/pull/57661#issuecomment-1543501710, it sounds like at least the _plan_ is to no longer allow modifying the global version of these variables in any module and only allow the scoped version for `zephyr_get` to be set -- if that is the plan and always going to be the case, then I have no problem with this approach; though, I still have my doubts about this policy being able to be consistently maintained, given that we have already seen the snippet implementation itself slip past that in spite of the `zephyr_get` infrastructure being already in place.

Anyway, it is the build system maintainer's call and I will make snippets use `EXTRA_DTC_OVERLAY_FILE` and `EXTRA_CONF_FILE` as discussed offline.

> Regarding Kconfigs/defconfigs, the reason I put the CMake variables under an `append:` was to be able to create sibling keys that would do things that have no direct CMake variables controlling them. Something like this:
> 
> ```yaml
> name: foo
> append:
>   EXTRA_DTC_OVERLAY_FILE: my.overlay
> defconfig: my_defconfig
> ```
> 
> Where we pick names like `defconfig`, etc. that should be reasonably self documenting about what their role is (and of course document them as well)

While I see no technical problem with that, the first question that comes to my mind is: "Why are Kconfig overlays handled through a CMake variable under `append` while defconfigs are not? What makes Kconfig overlays (or defconfigs, depending on how you look at it) special?"

For example:

```yaml
name: foo
append:
  I_AM_A_CMAKE_VAR: asdf
  ME_TOO: ghjk
  OVERLAY_CONFIG: my.conf
  YET_ANOTHER_CMAKE_VAR: zxcv
defconfig: my.defconfig
```

versus

```yaml
name: foo
cmake_vars:
  I_AM_A_CMAKE_VAR: asdf
  YOU_GET_THE_IDEA: ghjk
config_overlays:
- a.conf
- b.conf
defconfigs: 
- x.defconfig
- y.defconfig
```

I would say the latter looks more consistent and complete (as in that it does not expose the "under-the-hood" details like setting the CMake variables to specify the Kconfig overlays for the build system).

mbolivar-nordic commented:
> I would say the latter looks more consistent and complete (as in that it does not expose the "under-the-hood" details like setting the CMake variables to specify the Kconfig overlays for the build system).

The reason I didn't do it that way is because I think it is more self-documenting if the snippet file format explicitly mentions the build system variables that users are likely to be familiar with already.

tejlmand commented:
> Based on https://github.com/zephyrproject-rtos/zephyr/pull/57661#issuecomment-1543501710, it sounds like at least the plan is to no longer allow modifying the global version of these variables in any module and only allow the scoped version for zephyr_get to be set

Technically CMake doesn't have global scope variables, it has current scope and cache variables, but a variable set at the highest scope is available in all scopes below (although changes to the variable in child scopes will not affect (grand)-parent scopes).
(Though it is possible to adjust the immediate parent scope)
Current scope adjustment of variables will still be allowed, like here:
https://github.com/zephyrproject-rtos/zephyr/blob/2f718dd369e41a9ae50fb613ac29d8ce192da440/cmake/modules/zephyr_module.cmake#L81-L88
where Zephyr module defined settings are added in current scope, but settings added there will take lowest precedence.

We want the ability of defining the exact precedence for snippets, and therefore we need dedicated handling.

In case we wanted dedicated precedence handling of settings originating from Zephyr modules, then we should also create a dedicated scope for those.

tejlmand committed:
cmake: create Zephyr scope functions for variables

zephyr_get() supports multiple scopes when returning variables.
Variables are returned based on the scope's priority.

To facilitate creation of more scopes in Zephyr for variable handling
additional functions are introduced.
- zephyr_create_scope(<scope>)
    creates a new scope
- zephyr_set(<var> <val>... SCOPE <scope>)
    set the value of a variable in the specified scope.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
snippets: support for EXTRA_DTC_OVERLAY_FILE and EXTRA_CONF_FILE

This commit adds change snippets to use the new EXTRA_DTC_OVERLAY_FILE
and EXTRA_CONF_FILE Zephyr build system variables instead of
DTC_OVERLAY_FILE and OVERLAY_CONFIG.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@stephanosio and @mbolivar-nordic PR has been updated.

I added a new commit to support EXTRA_CONF_FILE and EXTRA_DTC_OVERLAY_FILE as #57994 has been merged.

For time being I kept support for DTC_OVERLAY_FILE and OVERLAY_CONFIG, but as the snippets feature is very new and has not yet been part of a release then I actually considered to remove support for DTC_OVERLAY_FILE  and OVERLAY_CONFIG completely in snippets and avoid any confusion and need of deprecation later.

Any thoughts / comments on this ?

tejlmand commented:
> For time being I kept support for DTC_OVERLAY_FILE and OVERLAY_CONFIG, but as the snippets feature is very new and has not yet been part of a release then I actually considered to remove support for DTC_OVERLAY_FILE and OVERLAY_CONFIG completely in snippets and avoid any confusion and need of deprecation later.
>
> Any thoughts / comments on this ?

Discussed offline with @mbolivar-nordic and have removed support for DTC_OVERLAY_FILE and OVERLAY_CONFIG in snippets.

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2602,10 +2613,73 @@ function(zephyr_get variable)
+# APPEND       : Append values to the already existing variable in <scope>
+#
+
+# Usage:
+#   zephyr_set(<variable> <value> SCOPE <scope> [APPEND])
+#
+# Zephyr extension of CMake set which allows a variable to be set in a specific
+# scope. The scope is used on later zephyr_get() invocation for precedence
+# handling when a variable it set in multiple scopes.
+#
+# <variable>   : Name of variable
+# <value>      : Value of variable, multiple values will create a list.
+#                The SCOPE argument identifies the end of value list.
+# SCOPE <scope>: Name of scope for the variable
+# APPEND       : Append values to the already existing variable in <scope>
```
duplicate

tejlmand replied:
fixed.

stephanosio reviewed (changes requested):
Looks good to me.

I can confirm that the introduction of `EXTRA_DTC_OVERLAY_FILE` fixed the problem with the `app.overlay` vs. snippet overlay issue.

There are still some outstanding issues (under investigation) causing the new snippet test to fail; but, they seem to be unrelated to the changes in this PR.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2602,10 +2613,73 @@ function(zephyr_get variable)
 
   if(GET_VAR_MERGE)
     list(REMOVE_DUPLICATES ${variable})
+    if(GET_VAR_REVERSE)
+      list(REVERSE ${variable})
+    endif()
```
> Won't these cause conflict? If there is a duplicate in the list, e.g. A, B, C, D, E, A it will then become A, B, C, D, E and reversed to E, D, C, B, A so A will be applied last when someone might want it to be applied first

This is exactly as designed.
https://github.com/zephyrproject-rtos/zephyr/blob/e8fab54b6605aa6b8432318777a2a582d4219842/cmake/modules/extensions.cmake#L2499-L2502

And the order of precedence means first in list has highest precedence, in your example A has higher precedence than B (and therefore duplicated As can be removed).
When you reverse the list, for example because what is applied last wins, such as `EXTRA_CONF_FILE` where last file applied wins, then you still want `A` to win over `B`, hence `REVERSE` can be applied.

> when someone might want it to be applied first

That someone should only consider in which scope they want to add their setting.
Sysbuild, cache, env, snippets, current, and the scope decides what takes precedence, and the consumer of `zephyr_get()` knows whether first in list means highest priority, or if last in list wins.

> although specifying it twice would not be recommended

Depends on the use-case.
If `A` is default applied by a sample (current scope), and user wants A to have higher precedence, then user can add A to the cache using `-D<var>=A`, and it move higher up the list (and the current scope version gets removed).
Whether the consumer of the list wants to reverse the list doesn't change the fact that the cache variant of A still has moved up the list.

stephanosio commented on cmake/modules/configuration_files.cmake:
```diff
@@ -116,5 +116,5 @@ DTC_OVERLAY_FILE=\"dts1.overlay dts2.overlay\"")
 
-zephyr_get(EXTRA_CONF_FILE SYSBUILD LOCAL VAR EXTRA_CONF_FILE OVERLAY_CONFIG)
-zephyr_get(EXTRA_DTC_OVERLAY_FILE SYSBUILD LOCAL)
+zephyr_get(EXTRA_CONF_FILE SYSBUILD LOCAL VAR EXTRA_CONF_FILE OVERLAY_CONFIG MERGE REVERSE)
+zephyr_get(EXTRA_DTC_OVERLAY_FILE SYSBUILD LOCAL MERGE REVERSE)
```
> However, for settings where the value last in the list will overwrite
values earlier in the list, we want the list to be reversed, examples
of such can be CONF_FILE, OVERLAY_CONFIG, DTC_OVERLAY_FILE, where the
content of the file last in the list will overrule the content from an
earlier file.

This seems to reverse the inner content from each scope, not just the order of the scopes, which I am pretty sure is not something desirable.

For example, with `-DEXTRA_CONF_FILE=a.conf;b.conf`, you would expect the `b.conf` to be processed after `a.conf`; but, the current implementation does the opposite.

tejlmand replied:
good point.
Taking a look.

tejlmand replied:
@stephanosio fixed.
